### PR TITLE
quantizr (0.33)

### PR DESCRIPTION
**Pull request by Quantizr**
# submission name:
quantizr

# upload zipped `archive.zip`
[archive.zip](https://github.com/user-attachments/files/26863641/archive.zip)

# report.txt
```
=== Evaluation config ===
  batch_size: 16
  device: cuda
  num_threads: 2
  prefetch_queue_depth: 4
  report: submissions/quantizr/report.txt
  seed: 1234
  submission_dir: submissions/quantizr
  uncompressed_dir: /mnt/d/comma_submission/comma_video_compression_challenge/videos
  video_names_file: /mnt/d/comma_submission/comma_video_compression_challenge/public_test_video_names.txt
=== Evaluation results over 600 samples ===
  Average PoseNet Distortion: 0.00051010
  Average SegNet Distortion: 0.00061113
  Submission file size: 299,970 bytes
  Original uncompressed size: 37,545,489 bytes
  Compression Rate: 0.00798951
  Final score: 100*segnet_dist + √(10*posenet_dist) + 25*rate = 0.33

```

# does your submission require gpu for evaluation (inflation)?
yes

# did you include the compression script? and want it to be merged?
yes

# additional comments
This challenge kind of nerd sniped me... but I think this is as much as I'm gonna work on this. Also I don't have a good name for this submission. The training script isn't the exact training script I used as I had three different scripts which I ran a few times finetuning the previous best checkpoint. I had an LLM combine the scripts and to make a 5 stage single file pipeline which mimics what I did but I haven't run the entire script to generate a new submission. It should be pretty close though. The training script is also very much not optimized and not performant and it takes a lot longer to train the model than it needs to.
I think sub 0.30 is possible just by sweeping through different conv dims but I don't think its worth the GPU time for me to do so. Further improvement could possibly be made with some neural mask encoding techniques instead of using ffmpeg but I'm not planning on spending more time on this.
Changes made from #53 include dropping optical flow and using Feature-wise Linear Modulation on pose vectors instead of using both masks. As a result the mask video only needs to encode half as many frames and can be done at a higher CRF. Normal convolutions were replaced with depthwise separable convs for parameter efficiency (model is now just 88k params and 64kb when compressed). 
This isn't new to this, but perhaps as a helpful hint to others the training code interpolates (up) the frames -> clamps/rounds -> interpolates (down) to match the evaluation code so that the gradients can compensate for resampling blur due to the aspect ratio changing slightly during interpolation and to compensate for uint8 rounding noise. The training code also shuffles the frame pairs which the original dataloader could not do. 

comparison_4x_new

Interesting to see that the compressed video almost looks like it has background hills and a daytime sky.

### CODE REVIEW

**Comment by YassineYousfi** [comment]
Very nice! @Quantizr 
I'm glad the challenge nerd sniped you :D Leave some time for the v2, v3, etc. 
It's also interesting to see the artifacts at the image boundary, due to padding in the conv layers I'm assuming

File: submissions/quantizr/compress.py [hunk 1]
```diff
@@ -0,0 +1,891 @@
...
+
+# -----------------------------
+# Path Resolution & Imports
+# -----------------------------
+# compress.py is in ./submissions/quantizr/, we need to access ./modules.py
+ROOT_DIR = Path(__file__).resolve().parent.parent.parent
+sys.path.insert(0, str(ROOT_DIR))
+
+from frame_utils import AVVideoDataset, DaliVideoDataset, segnet_model_input_size
+from modules import SegNet, PoseNet, DistortionNet, segnet_sd_path, posenet_sd_path
```
Comment (dllu) [inline submissions/quantizr/compress.py:35]: @YassineYousfi @Quantizr From the README.md, 

> External libraries and tools can be used and won't count towards compressed size, unless they use large artifacts (neural networks, meshes, point clouds, etc.), in which case those artifacts should be included in the archive and will count towards the compressed size. **This applies to the PoseNet and SegNet.**

Did you remember to add the SegNet and PoseNet weights to the compressed size? That should add about 90 MB to the final size.

Or do the rules only mean that you can't use SegNet and PoseNet in the inflation step (but you can use them in compression without adding to the size)?

**Review by dllu — COMMENTED** [review]
Question about the rules...

File: submissions/quantizr/compress.py (hunk 1 above)
Comment (YassineYousfi) [reply submissions/quantizr/compress.py:35]: Yes. They only count if they’re used in decompression. Next bullet point in the rules explains that.

**Comment by YassineYousfi** [comment]
@Quantizr This submission won an honorable prize. Please email me at {first name}@comma.ai for logistics.
Congratulations!
